### DISCUSSION TITLE
Streaming directly to S3?

### DISCUSSION THREAD

**Discussion by richdynamix** [awcodes/filament-curator#172] [Q&A] [answered]
Is it possible to setup Curator so that uploading media streams directly to S3? The reason for the request is to allow Curator to work within a Vapor deployed serverless environment? 

For reference, the image seems to make it to S3 from the admin, but when I click create media I get the following error - 

<img width="855" alt="Screenshot 2023-05-05 at 09 57 43" src="https://user-images.githubusercontent.com/1104549/236416918-fea520b8-5bb0-4850-a14a-cec17d85a7c5.png">

**Comment by awcodes** [comment]
I thought this was fixed previously. Might be a regression. I'll check. Thanks.

**Comment by richdynamix** [comment]
I can confirm the latest version fixes the upload issue. I can see the image in the S3 bucket and when I view the media item the image is there with the correct path. The only issues I am seeing now is the media grid view, and the calls in the frontend returning 404.

<img width="666" alt="Screenshot 2023-05-05 at 13 25 24" src="https://user-images.githubusercontent.com/1104549/236457052-06ade729-244e-49e9-9bc0-44c96a81cd6d.png">

<img width="862" alt="Screenshot 2023-05-05 at 13 37 32" src="https://user-images.githubusercontent.com/1104549/236459396-971d379e-c0a6-4334-8f2a-f24568d5eb04.png">

**Comment by awcodes** [comment]
Can you check the DB for me and see what the stored value for the disk is on the image?

**Reply by richdynamix** [reply]
@awcodes The disk is saved as S3. I did some digging in `vendor/awcodes/filament-curator/src/Curator.php` and the `getGlideServer()` uses the storage_path() helper for source and cache settings.

```php

    public function getGlideServer(): Server|ServerFactory
    {
        if (! $this->glideServer) {
            return ServerFactory::create([
                'driver' => $this->getGlideDriver(),
                'response' => new LaravelResponseFactory(app('request')),
                'source' => storage_path('app'),
                'source_path_prefix' => $this->getGlideSourcePathPrefix(),
                'cache' => storage_path('app'),
                'cache_path_prefix' => $this->getGlideCachePathPrefix(),
                'max_image_size' => $this->getGlideMaxImageSize(),
            ]);
        }

        return $this->glideServer;
    }
```

It doesn't look like this can load the file directly from S3 without some modifications to the Factory options. I followed the Glide docs by loading the disk driver directly - 

```php 

    public function getGlideServer(): Server|ServerFactory
    {
        if (! $this->glideServer) {
            return ServerFactory::create([
                'driver' => $this->getGlideDriver(),
                'response' => new LaravelResponseFactory(app('request')),
                'source' => Storage::disk('s3')->getDriver(),
                'cache' => Storage::disk('s3')->getDriver(),
                'cache_path_prefix' => $this->getGlideCachePathPrefix(),
                'max_image_size' => $this->getGlideMaxImageSize(),
            ]);
        }

        return $this->glideServer;
    }

```

This done the trick in my case, but a config option would be required for other disks to work as intended.

**Reply by awcodes** [reply, marked as answer]
Both of those can be set via the config using the facade.

**Reply by awcodes** [reply]
Good to know that was the issue though. Thanks.

**Reply by richdynamix** [reply]
@awcodes I have a working copy of the Beta V3 and while the README mentions `Custom Glide Server Factories will need to be updated to use the new GlideServerFactoryProvider` I cannot find any reference to the `GlideServerFactoryProvider` in the codebase except for the README.

I tried to implement my own ServerFactory and swap out the reference within the configuration file but it didn't seem to be used. I noticed the Trait `HasGlideSettings` within the codebase when looking for an alternative method to implement this but it looks like this Trait is no longer being used anywhere. Do you have any references or documentation on how I would implement a custom ServerFactory within V3?

**Reply by awcodes** [reply]
Sorry, I need to update the Readme in a lot of places. :)

Have a look at [https://github.com/awcodes/filament-curator/blob/3.x/src/Glide/DefaultServerFactory.php](https://github.com/awcodes/filament-curator/blob/3.x/src/Glide/DefaultServerFactory.php)
